src/compiler/program.ts

```diff
@@ -1537,21 +1537,25 @@ namespace ts {
                     const emitFilePath = toPath(emitFileName, currentDirectory, getCanonicalFileName);
                     // Report error if the output overwrites input file
                     if (filesByName.contains(emitFilePath)) {
-                        let chain: DiagnosticMessageChain;
-                        if (!options.configFilePath) {
-                            // The program is from either an inferred project or an external project
-                            chain = chainDiagnosticMessages(/*details*/ undefined, Diagnostics.Adding_a_tsconfig_json_file_will_help_organize_projects_that_contain_both_TypeScript_and_JavaScript_files_Learn_more_at_https_Colon_Slash_Slashaka_ms_Slashtsconfig);
+                        const sourceFile = filesByName.get(emitFilePath);
+                        if (options.noEmitOverwriteForJsFiles && isSourceFileJavaScript(sourceFile)) {

```

Comment:
actually that makes sense. That way you won't be surprised when opening a ts file along side a js file